File: api/src/main/java/net/blueshell/api/dto/response/AuthenticationDTO.java
package net.blueshell.api.dto.response;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import lombok.Data;
import lombok.EqualsAndHashCode;
import net.blueshell.api.base.BaseDTO;
import net.blueshell.api.common.enums.Role;

import java.io.Serial;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Set;

@Data
@EqualsAndHashCode(callSuper = true)
@Schema(name = "Login")
public class AuthenticationDTO extends BaseDTO {

    @Serial
    private static final long serialVersionUID = -8091879091924046844L;

    @NotBlank
    private final String token;

    @NotBlank
    private final long userId;

    @NotBlank
    private final String username;

    @NotBlank
    private final long expiration;

    @NotEmpty
    private final Set<Role> roles;

    public AuthenticationDTO(String token, long userId, String username, long expiration, Set<Role> roles) {
        this.token = token;
        this.userId = userId;
        this.username = username;
        this.expiration = expiration;
        this.roles = roles;
    }

    @JsonProperty("roles")
    public List<Role> getRolesSorted() {
        if (roles == null || roles.isEmpty()) return new ArrayList<>();

        return roles.stream()
                .sorted(Comparator.comparingInt(Enum::ordinal))
                .toList();
    }
}

File: api/src/main/kotlin/net/blueshell/api/dto/response/AuthenticationDTO.kt
package net.blueshell.api.dto.response

import com.fasterxml.jackson.annotation.JsonProperty
import io.swagger.v3.oas.annotations.media.Schema
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotEmpty
import net.blueshell.api.base.dto.BaseDTO
import net.blueshell.api.common.enums.Role
import java.io.Serial
import java.util.function.ToIntFunction

@Schema(name = "Login")
data class AuthenticationDTO(
    @field:NotBlank
    var token: String,

    @field:NotBlank
    var userId: Long,

    @field:NotBlank
    var username: String,

    @field:NotBlank
    var expiration: Long,

    @field:NotEmpty
    var roles: MutableSet<Role> = mutableSetOf(),

    var addressId: Long? = null
) : BaseDTO() {
    @get:JsonProperty("roles")
    val rolesSorted: MutableList<Role>
        get() {
            if (roles.isEmpty()) return ArrayList()

            return roles.stream()
                .sorted(Comparator.comparingInt(ToIntFunction { obj: Role -> obj.ordinal }))
                .toList()
        }

    companion object {
        @Serial
        val serialVersionUID = -8091879091924046844L
    }
}
